src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                    case 2:
+                        return
+                            // Compare ReturnType with the first argument
+                            SignatureComparer.Instance.HaveSameSignature(returnType, methodSymbol.Parameters.First(), caseSensitive: true) &&
+                            IsPredicate(methodSymbol.Parameters.Last().Type, typeArgument);
+
+                    default: return false;
+                }
+            }
+
+            private bool TryGetMethodSymbol(QueryClauseSyntax queryClause, out IMethodSymbol methodSymbol)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(queryClause, _cancellationToken);
+                methodSymbol = queryClauseInfo.OperationInfo.Symbol as IMethodSymbol;
+                return methodSymbol != null;
+            }
+
+            private static bool IsPredicate(ITypeSymbol typeSymbol, ITypeSymbol typeArgument)
+            {
+                if (typeSymbol.Name != "Func")  // TODO better that name check?
+                {
+                    return false;
+                }
+
+                var typeArguments = typeSymbol.GetTypeArguments();
+                if (typeArguments.Length != 2)
+                {
+                    return false;
+                }
+
+                return typeArguments.Last()?.SpecialType == SpecialType.System_Boolean &&
+                    SignatureComparer.Instance.HaveSameSignature(typeArguments.First(), typeArgument, caseSensitive: true);
+            }
+
+            private bool TryProcessFromClause(FromClauseSyntax fromClause, ProcessingState processingState)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(fromClause, _cancellationToken);
+                if (queryClauseInfo.OperationInfo.Symbol == null || // First FromClause may have no OperationInfo symbols.
+                    (queryClauseInfo.OperationInfo.Symbol is IMethodSymbol methodSymbol && methodSymbol.Parameters.Length == 2))
+                {
+                    processingState.AddVariableName(fromClause.Identifier.ValueText);
+                    processingState.AddStatementProcessingMethod(
+                        s => SyntaxFactory.ForEachStatement(
+                                    fromClause.Type ?? VarNameIdentifier,
+                                    fromClause.Identifier,
+                                    fromClause.Expression.WithoutTrivia(), WrapWithBlock(s)));
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool TryProcessLetClause(LetClauseSyntax letClause, ProcessingState processingState)
+            {
+                var queryClauseInfo = _semanticModel.GetQueryClauseInfo(letClause, _cancellationToken);
+                if (queryClauseInfo.OperationInfo.Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.Parameters.Length == 1)
```

Comment:
>methodSymbol.Parameters.Length == 1 [](start = 20, length = 35)

Why aren't we checking method's name?

Reply:
agreed to skip verification of query clauses for now

---
In reply to: [177200877](https://github.com/dotnet/roslyn/pull/25362#discussion_r177200877) [](ancestors = 177200877)